kubernetes-client/src/test/java/io/fabric8/kubernetes/client/dsl/internal/CustomResourceOperationsImplTest.java

```diff
@@ -0,0 +1,84 @@
+ * You may obtain a copy of the License at
+ *
+ *         http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.fabric8.kubernetes.client.dsl.internal;
+
+import java.io.IOException;
+
+import io.fabric8.kubernetes.api.model.KubernetesResource;
+import io.fabric8.kubernetes.api.model.apiextensions.CustomResourceDefinition;
+import io.fabric8.kubernetes.api.model.apiextensions.CustomResourceDefinitionBuilder;
+import io.fabric8.kubernetes.client.CustomResource;
+import io.fabric8.kubernetes.client.CustomResourceList;
+import io.fabric8.kubernetes.internal.KubernetesDeserializer;
+
+import org.junit.jupiter.api.Test;
+
+import com.fasterxml.jackson.core.JsonFactory;
+import com.fasterxml.jackson.core.JsonParser;
+import com.fasterxml.jackson.databind.MappingJsonFactory;
+
+import static org.hamcrest.core.IsInstanceOf.instanceOf;
+import static org.junit.Assert.assertThat;
+
+public class CustomResourceOperationsImplTest {
+
+	public static class MyCustomResource extends CustomResource {
+    }
+
+	public static class MyCustomResourceList extends CustomResourceList<MyCustomResource> {
+    }
+
+    CustomResourceDefinition crd = new CustomResourceDefinitionBuilder()
+        .withApiVersion("custom.group/v1alpha1")
+        .withKind("MyCustomResource")
+        .withNewMetadata()
+            .withName("custom.name")
+        .endMetadata()
+        .withNewSpec()
+            .withGroup("custom.group")
+            .withVersion("v1alpha1")
+            .withNewNames()
+                .withKind("MyCustomResource")
+                .withListKind("MyCustomResourceList")
+                .withPlural("mycustomresources")
+                .withSingular("mycustomresource")
+            .endNames()
+        .endSpec()
+    .build();
+
+    CustomResourceOperationContext context = new CustomResourceOperationContext()
+        .withCrd(crd)
+        .withType(MyCustomResource.class)
+        .withListType(MyCustomResourceList.class);
+
+    @Test
+	public void shouldRegisterWithKubernetesDeserializer() throws IOException {
+		// CustomResourceOperationsImpl constructor invokes KubernetesDeserializer::registerCustomKind
+		new CustomResourceOperationsImpl<>(context);
+
+		JsonFactory factory = new MappingJsonFactory();
+		JsonParser parser = factory.createParser("{\n" + 
+		    "    \"apiVersion\": \"custom.group/v1alpha1\",\n" + 
+		    "    \"kind\": \"MyCustomResource\"\n" + 
+		    "}");
+
+		KubernetesDeserializer deserializer = new KubernetesDeserializer();
+		KubernetesResource resource = deserializer.deserialize(parser, null);
+
+		assertThat(resource, instanceOf(MyCustomResource.class));
```

Comment:
I'd probably add an assertion validation the content of a field in order to verify deserialization worked.

(e.g. `assertThat(((MyCustomResource)resource).getApiVersion(), equalTo("custom.group/v1alpha1"))`)

Reply:
Added https://github.com/fabric8io/kubernetes-client/pull/2011/files#diff-1500b8672a6f6391abfa76c29b8128cfR83